Fix gosec issue:  Deferring unsafe Close() calls

ezrasilvera opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

When deferring a  `close()` calls on writable files, we don't have a chance to check the error returned from the `close()` call itself.  
Because the data might be cached in memory and not flushed to disk until closing the file we need to check the errors during `close()` to make sure the write() went OK.

Solution & Design:
For RW we treat the error as if it was a write error - we catch the error in a deferred function and update the error of the calling function. To do so we need to use named returned error variable in the calling function

For RO files, we catch the error and only add log messages 

To simplify the implementation we will use a single close function for both the RO and RW files. :  

`CloseIOAndCheckErr(c io.Closer, err *error)`
// CloseIOAndCheckErr closes the file and check the returned error. If there was an error a log messages will be printed.
// If a valid address (not nil) is passed in  err the function will also update the error


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Added comments also for opening RO files to in case someone will change to RW in the future

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
"NONE"
```

xpivarc commented:
/test pull-kubevirt-gosec

ezrasilvera commented:
/hold

ezrasilvera commented:
/unhold

ezrasilvera commented:
/retest

ezrasilvera commented:
/retest

danielBelenky commented on pkg/host-disk/host-disk.go:
```diff
@@ -88,18 +88,26 @@ func dirBytesAvailable(path string) (uint64, error) {
+	defer func() {
+		ferr := f.Close()
+		if err == nil {
+			err = ferr
```
This can hide some of the errors. This is probably an issue with Go's implementation of `file.Close`, but if the creation fails, you can't call close on it as you will get an `ErrInvalid` back.

One solution is to move this defer to line 105, after we ensure that the file was created.

ezrasilvera replied:
Which errors this hides?  I explicitly check to see if there are other errors and if so we don't care about the close() errors and I don't override the `err` variable

ezrasilvera replied:
Fixed

danielBelenky commented on pkg/host-disk/host-disk.go:
```diff
@@ -88,18 +88,26 @@ func dirBytesAvailable(path string) (uint64, error) {
 }
 
-func createSparseRaw(fullPath string, size int64) error {
+func createSparseRaw(fullPath string, size int64) (err error) {
```
Personally, I don't like the global error return. I think that it adds some "magic" to the code and makes it harder to read.

ezrasilvera replied:
This is not "global error return but rather named returned value which is a common pattern. 
On long functions this indeed might not be recommended but I think that for such short and clear function it's ok.

EdDev replied:
I think the named return values are mainly used for cases that the defer needs to set some value in them. I think the documentation also describes this pattern.
I saw another usage when initialization of the value is needed.. it may save a few lines of code.

But yes, it really confuses me when reading such code. I guess there is no choice but to get use to it when special cases like these are used.

ezrasilvera replied:
After a lot of investigations I found out that the **only way** to allow defer to change the returned value is to use named returned variables.  Even passing a pointer to the error will not help, I actually looked into the assembler to verify why this is the case . 
I will update the description of the issue so it is clear to  reviewers we must  use named retuned variables we want to check close(). So we can decide if it worth it...

danielBelenky commented on pkg/virt-launcher/virtwrap/util/libvirt_helper.go:
```diff
@@ -374,7 +375,7 @@ func SetupLibvirt() error {
 		return err
 	}
-	defer qemuConf.Close()
+	defer fileClose(qemuConf, &err)
```
You count here on `err` to be defined for you, but it might change and then we will hit a nil pointer here. Also, why wouldn't you just return the error from `fileClose`?

ezrasilvera replied:
Yes indeed I count on err be defined, but this is no different from any other place in all KubeVirt code where you take an address of variable someone defined earlier.  If someone changes the code he would need to take care of all implications.  If you think it would be better I can explicitly define and set it in the beginning .

The whole purpose of fileClose is to be called on `defer` you can return any value from it (or use that value). 
 I use that function to assign value to the  `err` variable which is then returned

EdDev commented on pkg/virt-handler/cache/cache.go:
```diff
@@ -221,12 +221,16 @@ func AddGhostRecord(namespace string, name string, socketFile string, uid types.
 			return err
 		}
 
-		defer f.Close()
```
Comparing the safety of the two, I find removing the `defer` more risky due to the potential maintenance issues with this function.

I suggest you extract this block to its own function and use defer there.

ezrasilvera replied:
I can certainly do that, although for such short function with single write operation adding the `close` into the err condition seems more clear.

EdDev replied:
Same as before, I think consistency is important.
In addition, future changes to the function may add another condition flow and there is a risk to forget about closing. The defer eliminates nicely such bugs. I think it is even more common to see it than the failure of a close call.

ezrasilvera replied:
See above comments - it's either explicit close() on each error condition or named returned variables.

EdDev commented on pkg/host-disk/host-disk.go:
```diff
@@ -88,18 +88,27 @@ func dirBytesAvailable(path string) (uint64, error) {
+		return
 	}
+
+	defer func() {
```
How about using a wrapper to do this?
`defer safeClose(f.close, &err)`

ezrasilvera replied:
We can use a wrapper as I did in libvirt_helper.go  , I personally think that a wrapper is more reasonable where there are multiple calls and for such single call I think inline function is fine. 
I will change to wrapper.

EdDev commented on pkg/host-disk/host-disk.go:
```diff
@@ -88,18 +88,27 @@ func dirBytesAvailable(path string) (uint64, error) {
 	_, err = f.WriteAt([]byte{0}, offset)
 	if err != nil {
-		return err
+		return
 	}
-	return nil
+	return
```
Do I miss something here or this is no longer needed now?

ezrasilvera replied:
Correct

EdDev commented on pkg/virt-launcher/virtwrap/util/libvirt_helper.go:
```diff
@@ -414,5 +415,12 @@ func SetupLibvirt() error {
+	return err // Need to use err and not nil to allow the defered func to set the value if needed
+}
+
+func fileClose(c io.Closer, err *error) {
```
I think you can generalize this by passing in a close function (`func() error`)

ezrasilvera replied:
Not sure I follow - can you elaborate what you propose ?

EdDev replied:
Sorry, the func name confused me. This is a generic close, not file specific. Maybe call it CloseWithErr or something similar.

EdDev commented on pkg/ephemeral-disk-utils/utils.go:
```diff
@@ -96,7 +96,7 @@ func Md5CheckSum(filePath string) ([]byte, error) {
 		return nil, err
 	}
-	defer file.Close()
+	defer file.Close() // #nosec no risk for readonly files
```
I always disliked these inline comments that come to satisfy an external tool.
It makes the code ugly, requiring me to read it when I walk through the code and distracts me.

There may be dozens on these all over now, forcing us to pollute the code with endless comments like these.
BTW, if someone adds a new defer like this tomorrow, or changes this one to a rw file instead. How will you track it?

ezrasilvera replied:
@EdDev I think the plan is to clean all gosec issues, and then add gosec to the regular build & PR tests so that a developer would be able to directly fix the gosec issues his code introduced. 
Regarding the comments.  I agree they might pollute the code, but I'm not sure what do you suggest should be the alterative. As I see it there are two option 1) disable a test globally 2) Review each potential violation and "ignore" it manually through this comment. 
Judging from the code I have already reviewed I think we should better go with (2) then disabling the test and then chasing possible (some times very hard) bugs.

ezrasilvera replied:
I agree this need to be generally discussed. Not only some of these are not handled in large projects , some are not even used in core libraries of golang itself !

ezrasilvera commented:
/hold

ezrasilvera commented:
/unhold

EdDev commented:
I think the topic itself is interesting.
I would have started doing this through a proposal to apply a policy on the project regarding `io.close()` as discussing it on a PR is less comfortable IMO.

> There are two ways we can check errors during close():
> 
>     1. Use explicit (not deferred) `close()` call that can be checked. Avoiding `defer` will require to close the file before every return which might be complex on long functions. We can create a utility function to check & close & return

The chance of a `close` to fail is lower than the chance of a coder missing to close it (especially when having to handler errors with early exit from the function).
It is also not going to handle cases where panic is raised (in such cases only `defer` will do the job).

Therefore, I think this option is a no-go.

> 
>     2. Use named returned error and a deferred function to close the file, and updated the named error if needed.

Seem like the only valid option to me.
But unlike what this PR is suggesting, I think we cannot ignore some cases with a comment.
If the `io.close()` function returns an error, we need to handle it always, no matter the case. If we do not need to handle it, it means it should not ever return an error or the linter needs to be fixed.

In addition, comments are not reliable. The reason for a condition to be safe may change immediately after and we will just ignore it.

If the decision is to enforce this handling, we should always do it, no matter what is the context.


Another point here is about what the caller will do with the error.
If the file was not closed, then do we expect the caller to explode? To check the error and retry?
If we do not know, I suggest we panic instead.

ezrasilvera commented:
@EdDev I think we kind of converged. 

I agree with you we should count on comments and should check Close() errors for both RO and RW files
So the only question we need to address is what do we do in case we detect a Close() error. I suggest the following:

- RW File: As the data might be flushed only on Close() n error during Close() may indicate that the data was not written correctly and therefore we should treat Close() errors the same way we treat a Write() error (Already included in the PR)
- RO file:  Close() is called after the needed information/data was already read successfully. So in term of the function the "task" succeeded and we shouldn't treat that as an error. For the RO case I suggest we log an error message so if the error has some side-effects it can help us analyze it and root-cause it.  

If you think further discussion is needed, we can either pull more people to this conversation or I can post the topic to the dev mailing list

ezrasilvera commented:
/retest

EdDev commented:
> I agree with you we should count on comments and should check Close() errors for both RO and RW files
> So the only question we need to address is what do we do in case we detect a Close() error. I suggest the following:
> 
>     * RW File: As the data might be flushed only on Close() n error during Close() may indicate that the data was not written correctly and therefore we should treat Close() errors the same way we treat a Write() error (Already included in the PR)

The question is what you expect the caller to do in this case.
Lets take the case that data is written to a file, it gets written and close fails.
Should the caller revert? Panic? Try to close again?

> 
>     * RO file:  Close() is called after the needed information/data was already read successfully. So in term of the function the "task" succeeded and we shouldn't treat that as an error. For the RO case I suggest we log an error message so if the error has some side-effects it can help us analyze it and root-cause it.

Failure to close a file is a failure, I do not see how we can take a decision to ignore it.
There are other consequences if a file cannot be closed, like leaking file descriptors.
 
> If you think further discussion is needed, we can either pull more people to this conversation or I can post the topic to the dev mailing list

I think it has implications on all the project and potentially other projects as well. Until now we either leaked fd/s or never had these at all.
I think it needs to be raised on a larger forum to see what others think about it.

ezrasilvera commented:
> 
> The question is what you expect the caller to do in this case.
> Lets take the case that data is written to a file, it gets written and close fails.
> Should the caller revert? Panic? Try to close again?
> 

What I meant is that we already handle this today: we return an error on Write() errors, and the caller handles those errors. From the caller perspective a Close() error is identical to Write() error, so there is no need for special behavior on the caller side and it's enough we will just return the error. 

The RO is a different case, as the actual `read` succeeded, so it's indeed interesting question if we should treat Close() as if we failed to read the file.

EdDev commented:
> What I meant is that we already handle this today: we return an error on Write() errors, and the caller handles those errors. From the caller perspective a Close() error is identical to Write() error, so there is no need for special behavior on the caller side and it's enough we will just return the error.

Ok, now I understand what you mean, thank you for the clarification.
I think there is a difference here: If "write" fails, one can assume nothing was written (not really sure how accurate this is), but if "close" fails the data may be already there and the caller may need to do something else.
  
> The RO is a different case, as the actual `read` succeeded, so it's indeed interesting question if we should treat Close() as if we failed to read the file.

Yes, here, depending on if the function has a side effect or not, the information may have been already consumed.

Seems like this "little" subject is very complex.
IMO we should just do the same thing no matter the context. Panic, log and ignore, log and return an error.
Personally the panic one sounds the most suitable here. Not being able to release a fd is a big thing, especially for services that live for a long time... this is pure leaking of resources. It will also allow us to detect if this is happening or not.
I'm interested to hear more opinions on this.

ezrasilvera commented:
I posted to the mailing list. 

> I think there is a difference here: If "write" fails, one can assume nothing was written (not really sure how accurate this is), but if "close" fails the data may be already there and the caller may need to do something else.
> 

I think it's identical - Write error doesn't say nothing was written so in both cases you may have partial writing or no writing

ezrasilvera commented:
/retest

ezrasilvera commented:
/retest

ezrasilvera commented:
@EdDev I believe the code is aligned with what we talked during the community call and with @davidvossel comments in the mailing list. 
RW files: handle the errors 9similar to write errors)
RO file: Log a message, but continue as usual (we don't have evidence that this is currently an existing issue)

EdDev commented on pkg/host-disk/host-disk.go:
```diff
@@ -88,18 +89,30 @@ func dirBytesAvailable(path string) (uint64, error) {
+	return err
+}
+
+func closeFileAndCheckErr(c io.Closer, err *error) {
```
You used `io.Closer` interface, so lets generalize it. How about `closeIOAndCheckErr` ?

And use `ioErr` or just `e` inside.

ezrasilvera replied:
Done

EdDev commented on pkg/virt-launcher/virtwrap/api/converter.go:
```diff
@@ -143,6 +143,7 @@ func checkDirectIOFlag(path string) bool {
 	if err != nil && !os.IsNotExist(err) {
 		return false
 	}
+	// #nosec No need to check Close() errors - no writing just checking O_DIRECT flag
```
Lets not use comments like these. Please log it.

ezrasilvera replied:
It's a missed left over.

ezrasilvera replied:
Fixed

EdDev commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -817,13 +817,14 @@ var updateHostsFile = func(entry string) error {
 	}
-	return nil
+
+	return file.Close()
```
I prefer you follow the same convention everywhere and use the helper functions.

ezrasilvera replied:
Done

EdDev reviewed (commented):
I prefer we have a single convention in the code on how to handle `Close()` on io.
AFAIU the middle ground is to log the read io and return error for write io.

Lets have two helpers that implement this under a package that serves all (e.g. `util/close.go`).

Future errors that are sourcing from gosec should use these helpers.
(we should mention the error there so devs will locate it easily)

WDYT?

ezrasilvera commented:
> 
> Lets have two helpers that implement this under a package that serves all (e.g. `util/close.go`).
> WDYT?
That was my intention but I didn't see many functions under `pkg/util`  So I didn't know if it's OK to add a file there (or functions into util.go") and this is the only common place ..

ezrasilvera commented on pkg/host-disk/host-disk.go:
```diff
@@ -88,18 +89,30 @@ func dirBytesAvailable(path string) (uint64, error) {
-func createSparseRaw(fullPath string, size int64) error {
+// Use named return err to allow checking Close() errors in defer
+func createSparseRaw(fullPath string, size int64) (err error) {
+	var f *os.File
```
Correct

ezrasilvera replied:
Fixed

ezrasilvera commented on pkg/virt-handler/cache/cache.go:
```diff
@@ -216,17 +219,17 @@ func AddGhostRecord(namespace string, name string, socketFile string, uid types.
 			return err
 		}
-		f, err := os.Create(recordPath)
+		f, err = os.Create(recordPath)
```
Fixed

ezrasilvera commented on pkg/virt-handler/cache/cache.go:
```diff
@@ -242,7 +245,14 @@ func AddGhostRecord(namespace string, name string, socketFile string, uid types.
 	}
 
-	return nil
+	return err
```
Fixed

ezrasilvera commented:
@EdDev  I updated the code, and I'm using now a general global utility function for all Close() cases. 
I believe all comments are now addressed.

EdDev commented on pkg/util/BUILD.bazel:
```diff
@@ -4,9 +4,13 @@ go_library(
     name = "go_default_library",
     srcs = [
         "hw_helper.go",
+        "os_utils.go",
```
As this is under the `util` folder, re-using the name internally does not add any valuable info.
Maybe just `os`? It is not a package name, so there should not be any namespace collision here.
Other options: `os_extender`, `os_extra`.

ezrasilvera commented:
/retest

EdDev commented on pkg/util/os_utils.go:
```diff
@@ -0,0 +1,38 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
```
nit: 2020

EdDev commented on pkg/util/os_utils.go:
```diff
@@ -0,0 +1,38 @@
+import (
+	"io"
+
+	"kubevirt.io/client-go/log"
```
Ahh I hoped to avoid depending on other libs. But I guess for logging all will need to use it anyway.

ezrasilvera commented:
/retest

EdDev commented on pkg/util/os_utils.go:
```diff
@@ -0,0 +1,38 @@
+	"kubevirt.io/client-go/log"
+)
+
+// If err=nil only print and don't update the error in the calling function
```
Can you add another line that starts with `CloseIOAndCheckErr` and continues with what it does?
Just to make it a standard [godoc convention](https://blog.golang.org/godoc).

EdDev reviewed (commented):
Thank you, looks great. There are some small comments left, let me know what you think.

EdDev commented:
Please remember to squash everything and add the info for the next generation in the commit message.

ezrasilvera commented:
> info for the next generation

?

EdDev commented:
> > info for the next generation
> 
> ?

Populate the commit message with the information the relates to this change: Short overview, intent and what changed. Basically I guess what you mentioned on the PR itself.
We need it in the git history.

EdDev commented on pkg/host-disk/host-disk.go:
```diff
@@ -88,13 +89,14 @@ func dirBytesAvailable(path string) (uint64, error) {
 }
 
-func createSparseRaw(fullPath string, size int64) error {
+// Use named return err to allow checking Close() errors in defer
```
This is new? I do not remember seeing this before.
Comments tend to rot. Unless used to explain something non-obvious from the code itself I prefer to avoid them. I think it should be obvious that if a named return value is given it is used for some purpose. It may be because of the close or other reasons.
We surely do not need to own or limit this to the defer close handling.

If you do not mind, I prefer these out.

ezrasilvera replied:
This comment is important so I prefer to leave it. If someone if the future decides to change the named variable to "regular" returned variable , (e.g., because "it's not pretty "), nothing obvious will break, it will just not work, and it will be very hard to track it down.

I will squash some of the similar commits.

EdDev replied:
You assume that they will look or trust the comment. Comments do not protect, type checking, reviews, linters and unit tests protect.

But I leave it to you and other reviewers to choose.

ezrasilvera replied:
Will remove from all.

EdDev reviewed (commented):
Thanks for all the changes.
Please consider dropping the comments you added, see the inline comment. Not a blocker for me.

I will wait for the squash of all the commits and the commit message for the lgtm. Thanks!

ezrasilvera commented:
/hold

ezrasilvera commented:
/unhold

ezrasilvera commented:
/retest

ezrasilvera commented:
/retest

xpivarc commented on pkg/util/os_helper.go:
```diff
@@ -0,0 +1,40 @@
+)
+
+// CloseIOAndCheckErr closes the file and check the returned error.
+// If there was an error a log messags will be printed.
```
messags -> messages

ezrasilvera replied:
Fixed

xpivarc commented on pkg/util/os_helper.go:
```diff
@@ -0,0 +1,40 @@
+// CloseIOAndCheckErr closes the file and check the returned error.
+// If there was an error a log messags will be printed.
+// If a valid address (not nil) is passed in  err the function will also update the error
+// Note: to update the error the calling funtion need to use named returned variable
```
If used as `defer` function

ezrasilvera replied:
Fixed description

xpivarc commented on pkg/virtctl/imageupload/imageupload.go:
```diff
@@ -210,7 +211,7 @@ func (c *command) run(cmd *cobra.Command, args []string) error {
 		return err
 	}
-	defer file.Close()
+	defer util.CloseIOAndCheckErr(file, nil)
```
Why do we choose to ignore the error here?

ezrasilvera replied:
The file is opened in RO mode. In this case we only log the error message.

xpivarc commented on pkg/virt-launcher/virtwrap/util/libvirt_helper.go:
```diff
@@ -340,8 +341,8 @@ func NewDomainFromName(name string, vmiUID types.UID) *api.Domain {
 
-func SetupLibvirt() error {
-
+// Use named return err to allow checking Close() errors in defer
```
We don't need this comment since `CloseIOAndCheckErr` has it.

ezrasilvera replied:
Fixed

dhiller reviewed (commented):
Thanks for your work! Some questions inline.

ezrasilvera commented on pkg/ephemeral-disk-utils/utils.go:
```diff
@@ -97,7 +98,7 @@ func Md5CheckSum(filePath string) ([]byte, error) {
 		return nil, err
 	}
-	defer file.Close()
+	defer util.CloseIOAndCheckErr(file, nil)
```
No. Please see the description in the PR. For RO files we only print log message, this was also discussed in the community meeting. passing `nil` means - "only log"

dhiller replied:
Thanks for the explanation!

ezrasilvera commented on pkg/util/os_helper.go:
```diff
@@ -0,0 +1,40 @@
+	if ferr := c.Close(); ferr != nil {
+		log.DefaultLogger().Reason(ferr).Error("Error when closing file")
+		// Update the calling error only in case there wasn't a different error already
+		if err != nil && *err == nil {
```
Not really. The 'Write()' operations are using synchronic error checking. So you get immediate error catch and return. The problem is that this deferred function will be called also in that case, so we want to preserve the  Write() error and not override with the close() error . In any case once an write/update/change error happen the close() is no longer relevant.

dhiller replied:
Thanks for the explanation, works for me :)

ezrasilvera commented on pkg/virt-launcher/virtwrap/util/cpu_utils.go:
```diff
@@ -33,7 +34,7 @@ func GetPodCPUSet() ([]int, error) {
 		return nil, err
 	}
-	defer file.Close()
+	defer util.CloseIOAndCheckErr(file, nil)
```
It is . It's RO file so we call the function with `nil` to only log potential error

ezrasilvera commented on pkg/virt-handler/cache/cache.go:
```diff
@@ -186,7 +187,8 @@ func HasGhostRecord(namespace string, name string) bool {
 }
 
-func AddGhostRecord(namespace string, name string, socketFile string, uid types.UID) error {
+// Use named return err to allow checking Close() errors in defer
```
Fixed

ezrasilvera commented:
@xpivarc @dhiller Thanks for your comments. I fixed them all

@EdDev I removed all the "// Use named return err to allow checking ......  "  from code

dhiller reviewed (approved):
/lgtm

EdDev reviewed (commented):
Thank you for removing the inline comments, it looks very clean now.

My only request is to squash the commits into one, I see no reason to have multiple commits here and the split itself makes little sense when looking in the history (some things even leak between commits).
For example, changes in bazel files are part of the changes we do, they have no history value to developer that tries to follow what has been done here.

Please note that in general, the commit message is preferred to be formatted per [this](https://chris.beams.io/posts/git-commit/). Most importantly the content better have title/subject with intent and the general solution description. I think the first commit looks nice.

ezrasilvera commented:
> Thank you for removing the inline comments, it looks very clean now.
> 
> My only request is to squash the commits into one, I see no reason to have multiple commits here and the split itself makes little sense when looking in the history (some things even leak between commits).

I actually squashed them ((the 16 I had)  ) into these 4, which I deliberately left. I do see a usage for differentiate this into separate commits. I already cheery pick one of them for some other work I'm doing.

EdDev commented:
> I actually squashed them ((the 16 I had) ) into these 4, which I deliberately left. I do see a usage for differentiate this into separate commits. I already cheery pick one of them for some other work I'm doing.

The commits I see now show how you developed the change and how it changed from the initial post. The commits I am looking for are logical changes that help explain the change or have a strong logical independence. This both helps the review process and future investigation of the history to understand the changes.

If I try to review this PR one commit at a time it is harder than looking at all changes as a whole. It is also hard to reason for each commit this way (in the commit message) but maybe you can improve it.

ezrasilvera commented:
> If I try to review this PR one commit at a time it is harder than looking at all changes as a whole. It is also hard to reason for each commit this way (in the commit message) but maybe you can improve it.

Let me try to improve the messages. 

I do find the commits history valuable for future assessment, as we went through different approaches and I wanted to maintain that history.  
I did remove all intermediate commits with typos, small changes, etc.. and left only the ones that have some merit on their own.

ezrasilvera commented:
/retst

EdDev commented:
> as we went through different approaches and I wanted to maintain that history.

Different approaches should not appear in the merged branch, there we should see only the end result of all the discussions in the review process. If someone needs such info, they should use the PR or the review tool the used.

ezrasilvera commented:
/retest

dhiller commented:
/approve
